src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs

```diff
@@ -16,382 +16,341 @@
+    private readonly FrozenDictionary<string, IFeatureInfo> _features;
+    private readonly IFeatureInfo[] _featureInfos;
+
+    private readonly ConcurrentDictionary<string, Lazy<IEnumerable<IFeatureInfo>>> _featureDependencies = new();
+    private readonly ConcurrentDictionary<string, Lazy<IEnumerable<IFeatureInfo>>> _dependentFeatures = new();
+
+    public ExtensionManager(
+        IApplicationContext applicationContext,
+        IEnumerable<IExtensionDependencyStrategy> extensionDependencyStrategies,
+        IEnumerable<IExtensionPriorityStrategy> extensionPriorityStrategies,
+        ITypeFeatureProvider typeFeatureProvider,
+        IFeaturesProvider featuresProvider,
+        ILogger<ExtensionManager> logger)
     {
-        private readonly IApplicationContext _applicationContext;
-
-        private readonly IExtensionDependencyStrategy[] _extensionDependencyStrategies;
-        private readonly IExtensionPriorityStrategy[] _extensionPriorityStrategies;
-        private readonly ITypeFeatureProvider _typeFeatureProvider;
-        private readonly IFeaturesProvider _featuresProvider;
-
-        private FrozenDictionary<string, ExtensionEntry> _extensions;
-        private List<IExtensionInfo> _extensionsInfos;
-        private FrozenDictionary<string, IFeatureInfo> _features;
-        private IFeatureInfo[] _featureInfos;
-
-        private readonly ConcurrentDictionary<string, Lazy<IEnumerable<IFeatureInfo>>> _featureDependencies = new();
-        private readonly ConcurrentDictionary<string, Lazy<IEnumerable<IFeatureInfo>>> _dependentFeatures = new();
-
-        private bool _isInitialized;
-        private readonly object _synLock = new();
-
-        public ExtensionManager(
-            IApplicationContext applicationContext,
-            IEnumerable<IExtensionDependencyStrategy> extensionDependencyStrategies,
-            IEnumerable<IExtensionPriorityStrategy> extensionPriorityStrategies,
-            ITypeFeatureProvider typeFeatureProvider,
-            IFeaturesProvider featuresProvider,
-            ILogger<ExtensionManager> logger)
-        {
-            _applicationContext = applicationContext;
-            _extensionDependencyStrategies = extensionDependencyStrategies as IExtensionDependencyStrategy[] ?? extensionDependencyStrategies.ToArray();
-            _extensionPriorityStrategies = extensionPriorityStrategies as IExtensionPriorityStrategy[] ?? extensionPriorityStrategies.ToArray();
-            _typeFeatureProvider = typeFeatureProvider;
-            _featuresProvider = featuresProvider;
-            L = logger;
-        }
+        L = logger;
 
-        public ILogger L { get; set; }
+        var modules = applicationContext.Application.Modules;
+        var loadedExtensions = new ConcurrentDictionary<string, ExtensionEntry>();
 
-        public IExtensionInfo GetExtension(string extensionId)
+        // Load all extensions in parallel
+        Parallel.ForEach(modules, (module, cancellationToken) =>
```

Comment:
Can we instead keep this logic in a method like `EnsureInitialized()` but call it in the constructor. You may want to called it `Initialize()`.

Also, converting the namespace scope to a file in a PR like this will make reviewing it very difficult. I suggest you revert the you change the namespace back so it's easy to review this PR. At some point in time, we'll convert all namespaces in the entire project. If this was a small class, I would not mind the namespace change

Reply:
Keeping the logic in a separate method would not allow to use `readonly` modifiers on the fields anymore. And I don't see how an `Initialize` method makes the code easier to follow.

Reply:
Wrapping the code in an Initialize method tells me that that code is being invoked is to initialize the class and will make it more readable and easier to follow. Placing so much code in the constructor makes the constructor very congested and hard to follow from maintenance perspective.

Reply:
I thought more about this and I'm going to change this back to an initialize method, but an explicit one. Doing so much work in the c'tor seems bad anyway.

Reply:
Hm, this is not going to be easy. Let me revert this change completely...